lightning/src/util/ser_macros.rs

```diff
@@ -372,7 +372,87 @@ macro_rules! impl_writeable_tlv_based {
 				}, {
 					$(($vectype, $vecfield)),*
 				});
-				_init_tlv_based_struct!({$($reqfield),*}, {$($field),*}, {$($vecfield),*})
+				_init_tlv_based_struct!($st, {$($reqfield),*}, {$($field),*}, {$($vecfield),*})
+			}
+		}
+	}
+}
+
+/// Implement Readable and Writeable for an enum, with struct variants stored as TLVs and tuple
+/// variants stored directly.
+/// The format is, for example
+/// impl_writeable_tlv_based_enum!(EnumName,
+///   (0, VariantName) => { (0, variant_field)}, {(1, variant_optional_field)}, {};
+///   (1, TupleVariant), (2, TupleVariantB),
+/// );
+/// The type is written as a single byte, followed by any variant data.
+/// Attempts to read an unknown type byte result in DecodeError::UnknownRequiredFeature.
+macro_rules! impl_writeable_tlv_based_enum {
+	($st: ident, $(($variant_id: expr, $variant_name: ident) =>
+		{$(($reqtype: expr, $reqfield: ident)),* $(,)*},
+		{$(($type: expr, $field: ident)),* $(,)*},
+		{$(($vectype: expr, $vecfield: ident)),* $(,)*}
+	),* $(,)*;
```

Comment:
The optional trailing separators are so that you can write the list with trailing commas, ie without them you wouldn't be able to say `(0, A) => {}, {}, {}, (2, B) => {}, {}, {}, }` which is very natural when they're on separate lines, not that its a big deal either way. I'm ok swapping `;` for `|` but I'm not sure that's really any more natural?

Reply:
Not suggesting to change the character. Just pointing out an example, but I found it a little odd that the three `{}` are not enclosed by another set of braces, which I think would allow using a comma instead? But maybe too verbose.

As discussed offline, an alternative we may consider is doing a syntax similar to protocol buffers, which would be more explicit, e.g.

```
optional foo = 1;
required bar = 2;
repeated baz = 4;
```

But no need to change in this PR if we'd prefer that.

Reply:
Hmm, yea, I think that would be cool, but then we'd want to do it for `write_tlv_fields!()` as well, which is a much larger change.